### DISCUSSION TITLE
Tilbakemeldinger fra første gjennomføring av kurset

### DISCUSSION THREAD

**Discussion by thomaswolff** [nrkno/dotnetskolen#7] [General]
## Utført fortløpende
- [x]  Kloning med Git URL for de som bruker HTTPS-autentisering fungerte ikke 
- [x] Copy-paste av kommandoer og kjøring på Linux/Mac fungerte ikke uten å endre `\` til `/`
- [x] Noen hadde problemer med open statement under punktet "Opprette EPG" i steg 5 i Rider. Fiks: `unload` og `reload` løsningen i Rider.
- [x] Vurdere å gi veiledning for å visualisere OpenAPI-kontrakten
- [x] Skrive notis om at man ikke kan kjøre testene før det er lagt til minst én test i testprosjektet
- [x] Ønske om peker til hvordan man kan sette opp UI for OpenAPI-dokumentasjonen
- [x] Ønske om beskrivelse av coverlet.coverage
- [x] I steg 10 så kaster testen som validerer opp mot JSON Schema exception fordi man returnerer `text dateAsString` istedenfor `json dateAsString`, og dermed feiler JSON Schema-valideringen. Dersom man hadde brukt `json dateAsString`, så ville ikke testen kastet exception, og heller JSON Schema-valideringen feilet. Det er muligens mindre forvirrende.

## Lavthengende frukt
- [x] Flytte steget med å lage HTML for ReDoc til steg 7, og så inkorporere det å eksponere den via API-et til steg 10
- [x] Bruke `async {}` og `let!` i integrasjonstester for bedre async-håndtering

## IDE-problemer
- [x] Folk har slitt litt med Rider etter å ha lagt til nye filer. Hjelper å unloade og reloade.
- [x] Noen har slitt med VS Code og Ionide etter man legger til Giraffe -> `dotnet build` og åpne filen på nytt.

## Flyt i veiledning
- [x] Ønsker å kjøre web-API-et tidligere slik at man ser hva man har laget. Typ. få `200 OK` på `/ping`. Flette utvikling av API og integrasjonstester for å ikke få så store bolker.
- [x] Steg 11 - gjøre om en og en type til DDD-style, og kjøre testene etter hver type, for å raskere se effekten av endringene

## Open source
- [ ] ~~Squashe alle commits, og force pushe~~ Fjerner heller lenker til NRK Slack. Det er ingen hemmeligheter i Git-historikken, og siden det uansett kommer til å bli endringer i fremtiden er det like greit å bare beholde Git-historikken slik den er.
- [x] Fjerne lenker til NRK Slack og andre interne ressurser
- [x] Si fra til folk at branchene deres blir offentlig med mindre de sletter dem
- [x] Legge til MIT-lisens

## Mer innføring i konsepter
- [x] Ønske om forklaring på de ulike Paket-filene (.paket/Paket.Restore.targets, paket.dependencies, paket.lock, */paket.references)
- [x] Motivasjon for Paket:
  - [x] Transitive avhengigheter
  - [ ] ~~Mangel på .lock-fil i NuGet, slik at byggserver tar den versjonen den selv ønsker~~ Dette stemmer ikke lengre: [https://devblogs.microsoft.com/nuget/enable-repeatable-package-restores-using-a-lock-file/](https://devblogs.microsoft.com/nuget/enable-repeatable-package-restores-using-a-lock-file/)
- [ ] Spørsmål om `.exe` og `.dll`-filer fungerer på Linux og Mac. 
  - [ ] https://docs.microsoft.com/en-us/dotnet/core/deploying/deploy-with-cli
  - [ ] https://medium.com/@v.cheshmi/clr-bcl-cil-and-assemblies-in-net-framework-3de8bf09e781
  - [ ] https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.NETCore.Platforms/src/runtime.json
  - [ ] Kan utvide veiledningen der hvor man kjører prosjektet første gangen, og vise hvordan man kan kjøre den kjørbare filen direkte etter prosjektet er kompilert.
    - [ ] Vise `Measure-Command { .src\api\bin\Debug\net5.0\NRK.Dotnetskolen.Api.exe }`
    - [ ] Vise `Measure-Command { dotnet run --project .\src\api\NRK.Dotnetskolen.Api.fsproj }`
    - [ ] Vise tilsvarende kommandoer som over for Linux. Da er det med `time` istedenfor `Measure-Command` 
    - [ ] https://github.com/nrkno/dotnetskolen/pull/21

## Nice to have
- [ ] Kan legge inn "checkpoints" i veiledningen hvor folk må gi et svar i chaten. På den måten kan man vite hvor langt folk har kommet.
- [ ] ~~Det er mulig å kjøre testene uten `Program`-modulen. Kan vurdere å sette opp test-prosjektene uten `Program`~~ Dette er mer en kuriositet, og for å følge enkelheten ved å bruke .NET CLI sine maler beholder vi det slik det er.

## Future
- [ ] Til ekstern workshop: kan sette opp et eget API som returnerer EPG-data, slik at folk kan integerere mot det istedenfor å ha in-memory database i `DataAccess.fs`
- [ ] Lage C#-spor

**Comment by heidisu** [comment]
I steg 6 på forskjellen på fact og theory så tenker jeg på det som at man kjører theory-testene en gang for hver parameter man gir inn, så i stedet for å skrive x fact-tester kan man ha en theory test med x inlinedata-annoteringer og resultatet blir x tester. (man vil teste samme funksjon med ulik input), kanskje det kunne komme tydeligere fram?

**Reply by thomaswolff** [reply]
Noe slikt? https://github.com/nrkno/dotnetskolen/commit/5e3ce2f58c1f30ab2b1cb87cf9817fa0cb91e8c4

**Comment by heidisu** [comment]
Liten typo i steg 10 `returnerer en DateTime option verdi som indkerer om det gikk bra eller ikke. parseAsDateTime`. indikerer mangler en i

**Reply by thomaswolff** [reply]
Wow, bra spotta! 🧐 Skal rette opp 👍
